### PR TITLE
Migrate JSBI source to TypeScript.

### PR DESCRIPTION
**Pull request by 12wrigja**
This is based off similar work I did for https://github.com/js-temporal/temporal-polyfill/pull/21.

Open questions:
 - how does changing "type" in package.json potentially affect JSBI users?
 - does the JSBI source need to be moved into a folder in order for the sourcemaps to end up correct in the npm package?
 - does the TS source need to be included in the package as well (via the package.json files list?)
 - is es2015 "low-enough" of a tsc emit before passing to babel? Or should the tsc emit be a much higher level emit and let babel do the bulk of the transpilation?

### CODE REVIEW

File: lib/jsbi.ts [hunk 1]
```diff
@@ -527,31 +526,31 @@ class JSBI extends Array {
...
 
   // Helpers.
 
-  static __zero() {
+  static __zero(): JSBI {
     return new JSBI(0, false);
   }
 
-  static __oneDigit(value, sign) {
+  static __oneDigit(value, sign: boolean): JSBI {
```
Comment (jakobkummerow) [inline lib/jsbi.ts:539]:
```suggestion
  static __oneDigit(value: number, sign: boolean): JSBI {
```

File: lib/jsbi.ts [hunk 2]
```diff
@@ -1084,14 +1083,14 @@ class JSBI extends Array {
...
-  static __comparisonResultToBool(result, op) {
+  static __comparisonResultToBool(result, op: 0|1|2|3) {
     switch (op) {
       case 0: return result < 0;
       case 1: return result <= 0;
       case 2: return result > 0;
       case 3: return result >= 0;
     }
-    throw new Error('unreachable');
+    // throw new Error('unreachable');
```
Comment (jakobkummerow) [inline lib/jsbi.ts:1093]: Just delete the line.

File: lib/jsbi.ts [hunk 3]
```diff
@@ -1104,6 +1103,7 @@ class JSBI extends Array {
         case 2: return x > y;
         case 3: return x >= y;
       }
+      throw new Error('unreachable');
```
Comment (jakobkummerow) [inline lib/jsbi.ts:1106]: Why is this needed? Can't you annotate {op} in this function as {0|1|2|3} too?

File: lib/jsbi.ts [hunk 4]
```diff
@@ -1761,12 +1762,13 @@ class JSBI extends Array {
     throw new TypeError('Cannot convert object to primitive value');
   }
 
-  static __toNumeric(value) {
+  // TODO input type?
```
Comment (jakobkummerow) [inline lib/jsbi.ts:1765]: any

File: lib/jsbi.ts [hunk 5]
```diff
@@ -1820,27 +1822,27 @@ class JSBI extends Array {
   }
 
   // Digit helpers.
-  __digit(i) {
+  __digit(i): number {
```
Comment (jakobkummerow) [inline lib/jsbi.ts:1825]:
```suggestion
  __digit(i: number): number {
```

Same for the next three functions below.

**Review by jakobkummerow — COMMENTED** [review]
Source changes look good, with a few minor comments.

What's with the (internal) functions that haven't received type annotations?

Hopefully Mathias can say more about the open questions.

Regarding target ES level: keep in mind that JSBI is squarely targeted at old browsers. We've had people specifically request IE11 compatibility fixes before. I'm not sure how they dealt with the fact that `class JSBI` is an ES2015 feature. So it might even make sense to target ES5. (For the majority of the code, it shouldn't make a difference anyway.)

**Comment by 12wrigja** [comment]
The problem with es5 transpilation is that the tsc output for `class extends Array` is broken when transpiling to es5. See [this TypeScript playground example](https://www.typescriptlang.org/play?target=1#code/MYGwhgzhAEBq0FMAeAXBA7AJjAggJzzAE9oBvAKGmkwHsAVACwEt0BzACgEozKrpga6CDRAIAdCBocARAAkEISdADuNPCEwBCaZwDcvAL7kj5USmgAjaAF5o6BMrhd9FsQAcArhAbtpAMRoaHRd3Lx9pACEwPGDycldaRhYOPSA) as a minimal repro.

File: lib/jsbi.ts (hunk 5 above)
Comment (12wrigja) [reply lib/jsbi.ts:1825]: Done.

File: lib/jsbi.ts (hunk 1 above)
Comment (12wrigja) [reply lib/jsbi.ts]: Done.

File: lib/jsbi.ts (hunk 3 above)
Comment (12wrigja) [reply lib/jsbi.ts]: Yes - Done.

**Comment by 12wrigja** [comment]
"What's with the (internal) functions that haven't received type annotations?" I had another branch with these and forgot to merge it in before pushing/reviewing. PTAL. I tried to infer the right types for things based on how they were used, but there's at least a couple places where the tsc control flow can't infer the right things, so some casts were necessary to retain the existing code.

**Comment by 12wrigja** [comment]
@justingrant You did a great job finding and helping me figure out the source-mapping issues with the Temporal TS migration - would you mind taking a look at JSBI as well?

**Review by jakobkummerow — APPROVED** [review]
Code changes LGTM. I'll leave the packaging questions to other reviewers.

File: .eslintrc.cjs [hunk 6]
```diff
@@ -18,17 +18,26 @@ module.exports = {
...
     'node': true
   },
+  'plugins': [
+    '@typescript-eslint'
+  ],
   'extends': [
     'eslint:recommended',
     'google'
   ],
+  "parser": "@typescript-eslint/parser",
```
Comment (mathiasbynens) [inline .eslintrc.cjs:28]: nit: here and below, please use single quotes around property names and for string literals, for consistency with the rest of the file

File: package.json [hunk 7]
```diff
@@ -1,5 +1,6 @@
 {
   "name": "jsbi",
+  "type": "module",
```
Comment (mathiasbynens) [inline package.json:3]: This is a breaking change. Consumers now have to `import` rather than `require()`. A more incremental step would be to stick to publishing CommonJS. Could we do that?

**Comment by mathiasbynens** [comment]
> how does changing "type" in package.json potentially affect JSBI users?

See inline comment. Can we avoid making this change?

> does the JSBI source need to be moved into a folder in order for the sourcemaps to end up correct in the npm package?

Hmm, I’m not sure… Could you test with DevTools?

> does the TS source need to be included in the package as well (via the `package.json` files list?)

No. The way I think about this is: the fact we’re using TS is an implementation detail, and does not influence the library that we end up publishing. We could include `.d.ts` typings in the package, however, since this is useful for developers.

> is es2015 "low-enough" of a tsc emit before passing to babel? Or should the tsc emit be a much higher level emit and let babel do the bulk of the transpilation?

The current approach seems fine to me.

**Review by mathiasbynens — CHANGES_REQUESTED** [review]

File: package.json (hunk 7 above)
Comment (12wrigja) [reply package.json:3]: The [docs](https://nodejs.org/api/packages.html#packages_type) seem to indicate this should only have effects on JS files within the JSBI project, and have no bearing on how the package is imported. There's also [instructions](https://nodejs.org/api/packages.html#packages_dual_package_hazard) that indicate how to make a package compatible with both ESM and CJS, and neither of the proposed solutions there indicate that they are a breaking change. We could probably use option 1?

That being said, I did make this change for Temporal, and it caused some issues for certain environments (see https://github.com/js-temporal/temporal-polyfill/issues/29), so if we didn't make that change here I'd revert it there and apply the same solution to both repositories.

We can avoid changing this field by renaming the transpilation output of tsc from .js to .mjs files (by changing the "build" script to be something like `"tsc && for f in tsc-out/*.js; do mv -- \"$f\" \"${f%.js}.mjs\"; done && rollup --config rollup.config.js",`, but it would be nice to avoid that if possible.

Comment (mathiasbynens) [reply package.json:3]: I think there's a misunderstanding. There are two ways of indicating a package contains JavaScript modules (as opposed to CommonJS):

1. using the `.mjs` extension
2. using `"type": "module"` (which treats all `.js` files as if they were `.mjs`)

You’re right that we could do either of those _if_ we wanted to export JSBI as JavaScript modules (as opposed to the status quo of exporting as CommonJS).

I think it’s just not necessary to make that change, at this point. Why do you want to export as modules? It’s orthogonal to migrating the source to TypeScript.

Ideally, migrating the source to TypeScript is an implementation detail that doesn’t affect what we export or how users should import the npm package.

Comment (mathiasbynens) [reply package.json:3]: Here’s what we’ve been publishing to npm until now:

```
$ npm pack --dry-run
npm notice 📦  jsbi@3.1.6
npm notice === Tarball Contents ===
npm notice 10.2kB LICENSE
npm notice 27.6kB dist/jsbi-cjs.js
npm notice 32.0kB dist/jsbi-umd.js
npm notice 1.2kB  package.json
npm notice 10.0kB README.md
npm notice 27.6kB dist/jsbi.mjs
npm notice 1.5kB  jsbi.d.ts
```

Ideally, after the TS migration, we still publish the same files with equivalent contents. (It would be nice to only publish the module version in the future, I agree, but it’s a separate change that shouldn’t block the TS migration IMHO.)

Comment (12wrigja) [reply package.json:3]: I think I better understand this now, especially after looking into https://github.com/js-temporal/temporal-polyfill/issues/29 and finding one potential fix (see https://github.com/js-temporal/temporal-polyfill/pull/35).

If we specify that this package uses modules, then we'd have to rename the existing non-module files in dist (so jsbi-cjs.js and jsbi-umd.js) to use a `.cjs` file extension, and while some tools can automatically deal with this others (and especially users loading this file by path on disk) might not (I'm not sure how the UMD bundle in particular is used by tools or humans directly).

I'll remove the key for now and work around the issues it causes within the build system, and file an issue to consider swapping this around in some breaking version of JSBI.

Comment (12wrigja) [reply package.json]: Done.

File: benchmarks/sub.mjs [hunk 8]
```diff
@@ -817,7 +817,7 @@ const dataSmall = [{
   r: '0x7bffcbe01'
 }];
 
-import JSBI from '../jsbi.mjs';
```
Comment (mathiasbynens) [inline benchmarks/sub.mjs:820]: It's annoying that TypeScript doesn't support this (I guess?), because it breaks `d8 benchmarks/sub.mjs` — we now have to always test the compiled output instead.

**Review by mathiasbynens — APPROVED** [review]

**Comment by mathiasbynens** [comment]
Released as v3.2.0.
